### PR TITLE
(PIE-51) Ability to define a metrics server to ship to

### PR DESCRIPTION
**Pull request by HelenCampbell**
This is a rebase of the work done in https://github.com/puppetlabs/puppetlabs-puppet_metrics_collector/pull/6 with the addition of a splunk_hec option.
All comments in the original PR should have hopefully been addressed.
Still currently a WIP.

### CODE REVIEW

**Comment by jarretlavallee** [comment]
@HelenCampbell Thanks for picking this up and making those changes! In #6 I was wrong about the `json2timeseriesdb` being in sync with `puppet-metrics-viewer`. The changes to it allowed sending the STDIN to the script from https://github.com/puppetlabs/puppet-metrics-viewer/pull/25.

File: manifests/pe_metric.pp [hunk 1]
```diff
@@ -38,10 +40,44 @@
...
+
+      $port_metrics_command = empty($metrics_server_port) ? {
+        false => "${local_metrics_command} --port ${metrics_server_port}",
+        true  => $local_metrics_command,
+      }
+
+      $metrics_command = $metrics_server_type ? {
+        'influxdb' => "${port_metrics_command} --influx-db ${metrics_server_db} > /dev/null",
+        'graphite' => "${port_metrics_command} > /dev/null",
+        'splunk_hec' => "${port_metrics_command} | puppet splunk_hec --sourcetype puppet:summary --saved_report > /dev/null",
```
Comment (mrzarquon) [inline manifests/pe_metric.pp]: should be the following for splunk_hec (no assumption on $path setting, etc) /opt/puppetlabs/bin/puppet splunk_hec --sourcetype puppet:metrics --pe_metrics

File: README.md [hunk 2]
```diff
@@ -57,6 +57,25 @@ Integer: How often to collect metrics, in minutes. Defaults to `5`.
...
 
+##### metrics_server_info
+
+Struct:
+  metrics_server_type => Enum['influxdb','graphite'],
+  hostname            => String,
+  port                => Optional[Integer],
+  db_name             => Optional[String],
+
+Specifies a metrics server to write data to. Currently it supports `influxdb` and `graphite` type servers. The parameters `metrics_server_type` and `hostname` are both required, and `dbname` is required for a `metrics_server_type` of `influxdb`.
```
Comment (mrzarquon) [inline README.md]: add splunk_hec as data source, link to splunk_hec module since users need that setup first and then just selection splunk_hec, etc.

**Comment by mrzarquon** [comment]
@jarretlavallee this is ready for your review / merge - we had to update puppetlabs/splunk_hec to 0.7.1, but now we have a way to enable splunk support OR the metrics / graphite servers, etc.

File: manifests/pe_metric.pp [hunk 3]
```diff
@@ -10,6 +10,9 @@
   String                    $metric_script_file = 'tk_metrics',
   Array[Hash]               $additional_metrics = [],
   Boolean                   $ssl                = true,
+  Boolean                   $use_splunk_hec     = false,
```
Comment (npwalker) [inline manifests/pe_metric.pp]: I'm kind of thinking we shouldn't special case splunk here.  What about having an `enum['graphite','influxdb','splunk'] export_to` parameter and get rid of metrics server type inside of metrics_server_info.

**Comment by HelenCampbell** [comment]
@npwalker I've done an initial attempt at pulling out the type and putting it into parameters. Still need to fully test but is this implementation along the lines of what you were thinking of? Thanks

**Comment by npwalker** [comment]
This seems good.  I wasn't necessarily thinking of moving all of the type out to individual parameters but that's mostly because I'm lazy and didn't want to have to add the parameters across all the classes. 

Since you already did the work to move them all out I think this is good and probably easier for most end-consumers vs using the hash of info.

**Pull request merged**